File: SYProgressHUD/SYProgressHUD.m
//
//  SYProgressHUD.m
//

#import "SYProgressHUD.h"

@interface SYProgressHUD ()

@end

@implementation SYProgressHUD

+ (instancetype)shareInstance
{
    static SYProgressHUD * hud = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        hud = [[SYProgressHUD alloc] init];
        hud.toastTime = 2.0;
    });
    return hud;
}

+ (void)SYProgressShowMsg:(NSString *)msg inView:(UIView *)view mode:(MBProgressHUDMode)mode customImg:(UIImageView *)imageView
{
    if ([SYProgressHUD shareInstance].hud != nil)
    {
        [[SYProgressHUD shareInstance].hud removeFromSuperview];
    }
    [SYProgressHUD shareInstance].hud = [MBProgressHUD showHUDAddedTo:view ? view : [UIApplication sharedApplication].delegate.window animated:YES];
    [SYProgressHUD shareInstance].hud.label.text = msg;
    [SYProgressHUD shareInstance].hud.label.numberOfLines = 0;
    [SYProgressHUD shareInstance].hud.mode = mode;
    if (mode == MBProgressHUDModeCustomView) {
        [SYProgressHUD shareInstance].hud.customView = imageView;
    }
    [SYProgressHUD shareInstance].hud.contentColor = [UIColor whiteColor];
    [SYProgressHUD shareInstance].hud.bezelView.style = MBProgressHUDBackgroundStyleSolidColor;
    [SYProgressHUD shareInstance].hud.bezelView.color = [UIColor blackColor];
    [SYProgressHUD shareInstance].hud.animationType = MBProgressHUDAnimationZoom;
    [SYProgressHUD shareInstance].hud.label.textColor = [UIColor whiteColor];
    [SYProgressHUD shareInstance].hud.removeFromSuperViewOnHide = YES;
    [SYProgressHUD shareInstance].hud.margin = 15;
}

+ (void)showMsg:(NSString *)msg
{
    [self SYProgressShowMsg:msg inView:nil mode:MBProgressHUDModeText customImg:nil];
    [[SYProgressHUD shareInstance].hud hideAnimated:YES afterDelay:[SYProgressHUD shareInstance].toastTime];
}

+ (void)showMsg:(NSString *)msg inView:(UIView *)view
{
    [self SYProgressShowMsg:msg inView:view mode:MBProgressHUDModeText customImg:nil];
    [[SYProgressHUD shareInstance].hud hideAnimated:YES afterDelay:[SYProgressHUD shareInstance].toastTime];
}

+ (void)showProgressWithMsg:(NSString *)msg
{
    [self SYProgressShowMsg:msg inView:nil mode:MBProgressHUDModeIndeterminate customImg:nil];
}

+ (void)showProgressWithMsg:(NSString *)msg inView:(UIView *)view
{
    [self SYProgressShowMsg:msg inView:view mode:MBProgressHUDModeIndeterminate customImg:nil];
}

+ (void)showSuccessWithMsg:(NSString *)msg;
{
    [self SYProgressShowMsg:msg inView:nil mode:MBProgressHUDModeCustomView customImg:[[UIImageView alloc] initWithImage:[[SYProgressHUD shareInstance] imageWithResourceName:@"success"]]];
    [[SYProgressHUD shareInstance].hud hideAnimated:YES afterDelay:[SYProgressHUD shareInstance].toastTime];
}

+ (void)showSuccessWithMsg:(NSString *)msg inView:(UIView *)view
{
    [self SYProgressShowMsg:msg inView:view mode:MBProgressHUDModeCustomView customImg:[[UIImageView alloc] initWithImage:[[SYProgressHUD shareInstance] imageWithResourceName:@"success"]]];
    [[SYProgressHUD shareInstance].hud hideAnimated:YES afterDelay:[SYProgressHUD shareInstance].toastTime];
}

+ (void)showFailureWithMsg:(NSString *)msg
{
    [self SYProgressShowMsg:msg inView:nil mode:MBProgressHUDModeCustomView customImg:[[UIImageView alloc] initWithImage:[[SYProgressHUD shareInstance] imageWithResourceName:@"fail"]]];
    [[SYProgressHUD shareInstance].hud hideAnimated:YES afterDelay:[SYProgressHUD shareInstance].toastTime];
}

+ (void)showFailureWithMsg:(NSString *)msg inView:(UIView *)view
{
    [self SYProgressShowMsg:msg inView:view mode:MBProgressHUDModeCustomView customImg:[[UIImageView alloc] initWithImage:[[SYProgressHUD shareInstance] imageWithResourceName:@"fail"]]];
    [[SYProgressHUD shareInstance].hud hideAnimated:YES afterDelay:[SYProgressHUD shareInstance].toastTime];
}

+ (void)showCircleProgressWithMsg:(NSString *)msg
{
    [self SYProgressShowMsg:msg inView:nil mode:MBProgressHUDModeAnnularDeterminate customImg:nil];
}

+ (void)showCircleProgressWithMsg:(NSString *)msg inView:(UIView *)view
{
    [self SYProgressShowMsg:msg inView:view mode:MBProgressHUDModeAnnularDeterminate customImg:nil];
}

- (void)setProgress:(CGFloat)progress
{
    _progress = progress;
    [SYProgressHUD shareInstance].hud.progress = progress;
}

+ (void)hidden
{
    [[SYProgressHUD shareInstance].hud hideAnimated:YES];
}

+ (void)hiddenDelay:(NSTimeInterval)delay
{
    [[SYProgressHUD shareInstance].hud hideAnimated:YES afterDelay:delay];
}

- (UIImage *)imageWithResourceName:(NSString *)name
{
    NSBundle * bundle = [NSBundle bundleForClass:[SYProgressHUD class]];
    NSURL * url = [bundle URLForResource:@"images" withExtension:@"bundle"];
    NSBundle * imageBundle = [NSBundle bundleWithURL:url];
    UIImage * image = [UIImage imageWithContentsOfFile:[imageBundle pathForResource:name ofType:@"png"]];
    return image;
}

@end

File: SYProgressHUD/Classes/SYProgressHUD.swift
//
//  SYProgressHUD.swift
//
//  基于MBProgressHUD的二次封装

import Foundation
import MBProgressHUD

public let HUD = SYProgressHUD.shared

public class SYProgressHUD {
    /// 单例
    public static let shared = SYProgressHUD()
    
    private var hud: MBProgressHUD?
    
    /// hud展示时间
    private var afterDelay: TimeInterval = 2
    
    /// hud背景颜色 默认黑色
    private var backgroundColor: UIColor = .black
    
    /// 配置hud属性
    public func config(afterDelay: TimeInterval? = nil,
                       backgroundColor: UIColor? = nil) {
        if afterDelay != nil { self.afterDelay = afterDelay! }
        if backgroundColor != nil { self.backgroundColor = backgroundColor! }
    }
    
    private func show(_ msg: String,
                      mode: MBProgressHUDMode? = .text,
                      view: UIView?,
                      image: UIImage? = nil,
                      progress: Float = 0,
                      autoHidden: Bool = true,
                      completion: (() -> Void)? = nil) {
        if hud != nil && mode != MBProgressHUDMode.annularDeterminate {
            hud?.removeFromSuperview()
            hud = nil
        }
        if hud == nil {
            hud = MBProgressHUD.showAdded(to: view ?? UIApplication.shared.windows.last!,
                                          animated: true)
        }
        hud?.label.numberOfLines = 0
        hud?.contentColor = .white
        hud?.bezelView.style = .solidColor
        hud?.bezelView.color = self.backgroundColor
        hud?.animationType = .zoom
        hud?.margin = 15
        hud?.removeFromSuperViewOnHide = true
        hud?.minSize = CGSize(width: 80, height: 50)
        hud?.label.text = msg
        hud?.mode = mode!
        if mode == .annularDeterminate { hud?.progress = progress }
        if let image = image { hud?.customView = UIImageView(image: image) }
        if autoHidden { hud?.hide(animated: true, afterDelay: self.afterDelay) }
        if completion != nil {
            DispatchQueue.main.asyncAfter(deadline: .now() + self.afterDelay) {
                completion!()
            }
        }
    }
    
    private func hide() {
        hud?.hide(animated: true)
    }
}

extension SYProgressHUD {
    /// 文字hud
    public func showMsg(_ msg: String,
                        view: UIView? = nil,
                        completion: (() -> Void)? = nil) {
        show(msg, view: view)
    }
    
    /// 菊花hud
    public func showIndicator(_ msg: String,
                              view: UIView? = nil,
                              completion: (() -> Void)? = nil) {
        show(msg, mode: .indeterminate, view: view, autoHidden: false, completion: completion)
    }
    
    /// 成功hud
    public func showSuccess(_ msg: String,
                            view: UIView? = nil,
                            completion: (() -> Void)? = nil) {
        let path = Bundle(for: SYProgressHUD.self).path(forResource: "SYProgressHUD", ofType: "bundle")
        let bundle = Bundle(path: path!)
        let file = bundle?.path(forResource: "success@2x", ofType: "png")
        let image = UIImage(contentsOfFile: file!)?.withRenderingMode(.alwaysOriginal)
        show(msg, mode: .customView ,view: view, image: image, completion: completion)
    }
    
    /// 失败hud
    public func showFailed(_ msg: String,
                           view: UIView? = nil,
                           completion: (() -> Void)? = nil) {
        let path = Bundle(for: SYProgressHUD.self).path(forResource: "SYProgressHUD", ofType: "bundle")
        let bundle = Bundle(path: path!)
        let file = bundle?.path(forResource: "fail@2x", ofType: "png")
        let image = UIImage(contentsOfFile: file!)?.withRenderingMode(.alwaysOriginal)
        show(msg, mode: .customView ,view: view, image: image, completion: completion)
    }
    
    /// 进度hud
    public func showProgress(_ msg: String,
                             progress: Float,
                             view: UIView? = nil,
                             completion: (() -> Void)? = nil) {
        show(msg, mode: .annularDeterminate, view: view, progress: progress, autoHidden: false, completion: completion)
    }
    
    /// 隐藏hud
    public func hidden() {
        hide()
    }
}

